File: market/orderbook.go
package market

import (
	"errors"
	"sort"
)

// Side represents the direction of an order.
type Side int

const (
	Buy Side = iota
	Sell
)

// Order is a resting or incoming order in the book.
// AgentID links back to whichever agent (retail, fund, market maker, etc.)
// placed it — the book itself doesn't care what kind of agent that is.
type Order struct {
	ID       uint64
	AgentID  string
	Side     Side
	Price    float64 // ignored for market orders
	Quantity float64
	IsMarket bool
}

// Fill represents one match between an incoming order and a resting order.
type Fill struct {
	TakerAgentID string
	MakerAgentID string
	Price        float64
	Quantity     float64
}

// priceLevel holds all resting quantity at a single price, FIFO by arrival.
type priceLevel struct {
	price  float64
	orders []*Order // queue, front = oldest = filled first
}

// OrderBook is a lightweight limit order book: discrete price levels,
// each holding a FIFO queue of resting orders. This is intentionally
// simple (no full L3 depth) — enough to get realistic slippage and
// price-impact behavior without the complexity of a production matching engine.
type OrderBook struct {
	bids        []*priceLevel // sorted descending by price (best bid first)
	asks        []*priceLevel // sorted ascending by price (best ask first)
	nextOrderID uint64
}

func NewOrderBook() *OrderBook {
	return &OrderBook{}
}

var ErrEmptyBook = errors.New("no liquidity available on the requested side")

// BestBid returns the highest resting buy price, or (0, false) if none.
func (b *OrderBook) BestBid() (float64, bool) {
	if len(b.bids) == 0 {
		return 0, false
	}
	return b.bids[0].price, true
}

// BestAsk returns the lowest resting sell price, or (0, false) if none.
func (b *OrderBook) BestAsk() (float64, bool) {
	if len(b.asks) == 0 {
		return 0, false
	}
	return b.asks[0].price, true
}

// Spread returns ask-bid, or (0, false) if either side is empty.
func (b *OrderBook) Spread() (float64, bool) {
	bid, ok1 := b.BestBid()
	ask, ok2 := b.BestAsk()
	if !ok1 || !ok2 {
		return 0, false
	}
	return ask - bid, true
}

// MidPrice returns (bid+ask)/2, or (0, false) if either side is empty.
func (b *OrderBook) MidPrice() (float64, bool) {
	bid, ok1 := b.BestBid()
	ask, ok2 := b.BestAsk()
	if !ok1 || !ok2 {
		return 0, false
	}
	return (bid + ask) / 2, true
}

// AddLimitOrder places a resting limit order on the book without matching.
// Use Submit() for orders that should attempt to match immediately —
// this is exposed separately mainly for seeding initial liquidity
// (e.g. a market maker populating both sides at the start of a tick).
func (b *OrderBook) AddLimitOrder(o *Order) {
	b.nextOrderID++
	o.ID = b.nextOrderID
	if o.Side == Buy {
		b.bids = insertLevel(b.bids, o, true)
	} else {
		b.asks = insertLevel(b.asks, o, false)
	}
}

func insertLevel(levels []*priceLevel, o *Order, descending bool) []*priceLevel {
	for _, lvl := range levels {
		if lvl.price == o.Price {
			lvl.orders = append(lvl.orders, o)
			return levels
		}
	}
	levels = append(levels, &priceLevel{price: o.Price, orders: []*Order{o}})
	sort.Slice(levels, func(i, j int) bool {
		if descending {
			return levels[i].price > levels[j].price
		}
		return levels[i].price < levels[j].price
	})
	return levels
}

// Submit processes an incoming order (market or aggressive limit),
// walking the opposite side of the book and returning every fill
// generated. This is where slippage and price impact come from directly:
// a large order consumes multiple price levels, so the average fill
// price gets worse the deeper it has to walk — no separate formula needed.
//
// Any unfilled remainder of a limit order is added to the book as
// resting liquidity. Unfilled remainder of a market order is dropped
// (in a real market this is roughly "fill or kill" behavior for the
// portion that couldn't be matched; adjust here if you want your sim
// to model failed/partial market orders differently).
func (b *OrderBook) Submit(o *Order) []Fill {
	var fills []Fill
	remaining := o.Quantity

	opposite := &b.asks
	if o.Side == Sell {
		opposite = &b.bids
	}

	for remaining > 0 && len(*opposite) > 0 {
		lvl := (*opposite)[0]

		// price limit check: for non-market orders, stop if the level
		// no longer satisfies the limit price.
		if !o.IsMarket {
			if o.Side == Buy && lvl.price > o.Price {
				break
			}
			if o.Side == Sell && lvl.price < o.Price {
				break
			}
		}

		for len(lvl.orders) > 0 && remaining > 0 {
			resting := lvl.orders[0]
			qty := min(remaining, resting.Quantity)

			fills = append(fills, Fill{
				TakerAgentID: o.AgentID,
				MakerAgentID: resting.AgentID,
				Price:        lvl.price,
				Quantity:     qty,
			})

			resting.Quantity -= qty
			remaining -= qty

			if resting.Quantity <= 0 {
				lvl.orders = lvl.orders[1:]
			}
		}

		if len(lvl.orders) == 0 {
			*opposite = (*opposite)[1:]
		}
	}

	// resting remainder for limit orders only
	if remaining > 0 && !o.IsMarket {
		o.Quantity = remaining
		b.AddLimitOrder(o)
	}

	return fills
}

func min(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

// DepthAtLevels returns cumulative quantity available within the given
// number of price levels on each side — useful for agents that want to
// gauge liquidity before sizing an order (e.g. a hedge fund checking
// whether the book can absorb its intended trade without excessive slippage).
func (b *OrderBook) DepthAtLevels(n int) (bidQty, askQty float64) {
	for i := 0; i < n && i < len(b.bids); i++ {
		for _, o := range b.bids[i].orders {
			bidQty += o.Quantity
		}
	}
	for i := 0; i < n && i < len(b.asks); i++ {
		for _, o := range b.asks[i].orders {
			askQty += o.Quantity
		}
	}
	return bidQty, askQty
}

File: src/market/orderbook.rs
use serde::{Deserialize, Serialize};

/// Side represents the direction of an order.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Side {
    Buy,
    Sell,
}

impl std::fmt::Display for Side {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Side::Buy => write!(f, "BUY"),
            Side::Sell => write!(f, "SELL"),
        }
    }
}

/// Order is a resting or incoming order in the book.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Order {
    pub id: u64,
    pub agent_id: String,
    pub side: Side,
    pub price: f64, // ignored for market orders
    pub quantity: f64,
    pub is_market: bool,
}

/// Fill represents one match between an incoming order and a resting order.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Fill {
    pub taker_agent_id: String,
    pub maker_agent_id: String,
    pub price: f64,
    pub quantity: f64,
}

/// PriceLevel holds all resting quantity at a single price, FIFO by arrival.
#[derive(Debug, Clone)]
pub struct PriceLevel {
    pub price: f64,
    pub orders: Vec<Order>, // queue, front = oldest = filled first
}

/// OrderBook is a lightweight limit order book: discrete price levels,
/// each holding a FIFO queue of resting orders.
#[derive(Debug, Clone, Default)]
pub struct OrderBook {
    pub bids: Vec<PriceLevel>, // sorted descending by price (best bid first)
    pub asks: Vec<PriceLevel>, // sorted ascending by price (best ask first)
    next_order_id: u64,
}

/// BookLevel summarizes price, cumulative quantity, and order count at a single price tier.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BookLevel {
    pub price: f64,
    pub quantity: f64,
    pub orders_count: usize,
}

impl OrderBook {
    pub fn new() -> Self {
        Self::default()
    }

    /// Best bid returns the highest resting buy price, or None if none.
    pub fn best_bid(&self) -> Option<f64> {
        self.bids.first().map(|lvl| lvl.price)
    }

    /// Best ask returns the lowest resting sell price, or None if none.
    pub fn best_ask(&self) -> Option<f64> {
        self.asks.first().map(|lvl| lvl.price)
    }

    /// Spread returns ask - bid, or None if either side is empty.
    pub fn spread(&self) -> Option<f64> {
        match (self.best_bid(), self.best_ask()) {
            (Some(bid), Some(ask)) => Some(ask - bid),
            _ => None,
        }
    }

    /// MidPrice returns (bid + ask) / 2, or if only one side has liquidity, returns that side's best price.
    pub fn mid_price(&self) -> Option<f64> {
        match (self.best_bid(), self.best_ask()) {
            (Some(bid), Some(ask)) => Some((bid + ask) / 2.0),
            (Some(bid), None) => Some(bid),
            (None, Some(ask)) => Some(ask),
            (None, None) => None,
        }
    }

    /// AddLimitOrder places a resting limit order on the book without matching.
    pub fn add_limit_order(&mut self, mut o: Order) -> u64 {
        self.next_order_id += 1;
        o.id = self.next_order_id;
        let side = o.side;
        if side == Side::Buy {
            insert_level(&mut self.bids, o, true);
        } else {
            insert_level(&mut self.asks, o, false);
        }
        self.next_order_id
    }

    /// CancelAgentOrders removes all unexecuted resting orders belonging to agent_id from both sides.
    pub fn cancel_agent_orders(&mut self, agent_id: &str) {
        self.bids = filter_levels(std::mem::take(&mut self.bids), agent_id);
        self.asks = filter_levels(std::mem::take(&mut self.asks), agent_id);
    }

    /// Submit processes an incoming order (market or aggressive limit),
    /// walking the opposite side of the book and returning every fill generated.
    pub fn submit(&mut self, mut o: Order) -> Vec<Fill> {
        let mut fills = Vec::new();
        let mut remaining = o.quantity;

        let is_buy = o.side == Side::Buy;
        let opposite = if is_buy {
            &mut self.asks
        } else {
            &mut self.bids
        };

        while remaining > 0.0 && !opposite.is_empty() {
            let lvl_price = opposite[0].price;

            // Price limit check for non-market orders
            if !o.is_market {
                if is_buy && lvl_price > o.price {
                    break;
                }
                if !is_buy && lvl_price < o.price {
                    break;
                }
            }

            let lvl = &mut opposite[0];
            while !lvl.orders.is_empty() && remaining > 0.0 {
                let resting = &mut lvl.orders[0];
                let qty = remaining.min(resting.quantity);

                fills.push(Fill {
                    taker_agent_id: o.agent_id.clone(),
                    maker_agent_id: resting.agent_id.clone(),
                    price: lvl_price,
                    quantity: qty,
                });

                resting.quantity -= qty;
                remaining -= qty;

                if resting.quantity <= 1e-9 {
                    lvl.orders.remove(0);
                }
            }

            if opposite[0].orders.is_empty() {
                opposite.remove(0);
            }
        }

        // Resting remainder for limit orders only
        if remaining > 1e-9 && !o.is_market {
            o.quantity = remaining;
            self.add_limit_order(o);
        }

        fills
    }

    /// DepthAtLevels returns cumulative quantity available within the given
    /// number of price levels on each side.
    pub fn depth_at_levels(&self, n: usize) -> (f64, f64) {
        let bid_qty: f64 = self
            .bids
            .iter()
            .take(n)
            .flat_map(|lvl| &lvl.orders)
            .map(|o| o.quantity)
            .sum();

        let ask_qty: f64 = self
            .asks
            .iter()
            .take(n)
            .flat_map(|lvl| &lvl.orders)
            .map(|o| o.quantity)
            .sum();

        (bid_qty, ask_qty)
    }

    /// TopLevels returns the top n resting price levels for bids (highest first) and asks (lowest first).
    pub fn top_levels(&self, n: usize) -> (Vec<BookLevel>, Vec<BookLevel>) {
        let n = if n == 0 { 15 } else { n };

        let bids = self
            .bids
            .iter()
            .take(n)
            .map(|lvl| BookLevel {
                price: lvl.price,
                quantity: lvl.orders.iter().map(|o| o.quantity).sum(),
                orders_count: lvl.orders.len(),
            })
            .collect();

        let asks = self
            .asks
            .iter()
            .take(n)
            .map(|lvl| BookLevel {
                price: lvl.price,
                quantity: lvl.orders.iter().map(|o| o.quantity).sum(),
                orders_count: lvl.orders.len(),
            })
            .collect();

        (bids, asks)
    }
}

fn filter_levels(levels: Vec<PriceLevel>, agent_id: &str) -> Vec<PriceLevel> {
    let mut remaining = Vec::with_capacity(levels.len());
    for mut lvl in levels {
        lvl.orders.retain(|o| o.agent_id != agent_id);
        if !lvl.orders.is_empty() {
            remaining.push(lvl);
        }
    }
    remaining
}

fn insert_level(levels: &mut Vec<PriceLevel>, o: Order, descending: bool) {
    for lvl in levels.iter_mut() {
        if (lvl.price - o.price).abs() < 1e-9 {
            lvl.orders.push(o);
            return;
        }
    }

    levels.push(PriceLevel {
        price: o.price,
        orders: vec![o],
    });

    if descending {
        levels.sort_by(|a, b| b.price.partial_cmp(&a.price).unwrap_or(std::cmp::Ordering::Equal));
    } else {
        levels.sort_by(|a, b| a.price.partial_cmp(&b.price).unwrap_or(std::cmp::Ordering::Equal));
    }
}
